comm/src/main/scala/coop/rchain/comm/network.scala

```diff
@@ -49,36 +49,41 @@ case class UnicastNetwork(id: NodeIdentifier, endpoint: Endpoint) extends Protoc
       table.observe(new ProtocolNode(sender, this))
       msg match {
         case ping @ PingMessage(_, _)     => handlePing(sender, ping)
-        case pong @ PongMessage(_, _)     => handlePong(sender, pong)
         case lookup @ LookupMessage(_, _) => handleLookup(sender, lookup)
-        case lookupResponse @ LookupResponseMessage(_, _) =>
-          handleLookupResponse(sender, lookupResponse)
+        case resp: ProtocolResponse => handleResponse(sender, resp)
       }
     }
 
   /**
-    * Validate incoming PING and send responding PONG.
+    *
     */
-  private def handlePing(sender: PeerNode, ping: PingMessage): Unit =
+  private def handleResponse(sender: PeerNode, msg: ProtocolResponse): Unit =
     for {
-      pong <- ping.response(local)
+      ret <- msg.returnHeader
+      promise <- pending.get(PendingKey(sender.key, ret.timestamp, ret.seq))
     } {
-      comm.send(pong.toByteSeq, sender)
+      try {
+        promise.success(Right(msg))
+      } catch {
+        case ex: java.lang.IllegalStateException => () // Future already completed
```

Comment:
Is handling a response twice expected to happen?  If not, we should log this or at least add a TODO to add logging later.  If it's expected to happen often, that seems like an inefficiency and we should try to fix it.

Reply:
It is not expected to happen, but
1. it's possible to double receive a udp packet, and there's an opportunity between the first completion of the promise and the removal of the promise to get the second received packet; and,
1. we don't have control over the sender of the message (imagine the first thirty Rust and Go programs that start getting run and debugged).